crates/bin/signer/src/main.rs

```diff
@@ -111,34 +112,113 @@ impl signer::Signer for SignerState {
         &self,
         verify_proofs_request: Request<VerifyProofsRequest>,
     ) -> Result<Response<VerifyProofsResponse>, Status> {
-        for proof in &verify_proofs_request.get_ref().proofs {
-            let keyset_id = KeysetId::from_bytes(&proof.keyset_id)
-                .map_err(|_| Status::invalid_argument(Error::BadKeysetId))?;
+        for proof in &verify_proofs_request.get_ref().proofs {                  
+            let mut err_details = ErrorDetails::new(); 
+            let keyset_id = match KeysetId::from_bytes(&proof.keyset_id) {
+                Ok(keyset_id) => keyset_id,
+                Err(_) => {
+                    err_details.add_bad_request_violation(
+                        "keyset_id".to_string(),
+                        "error when parsing keyset_id from bytes".to_string()
+                    );
+                    let status = Status::with_error_details( 
+                        Code::FailedPrecondition,
+                        "error when parsing keyset_id from bytes".to_string(),
+                        err_details
+                    );
+                    return Err(status);
+                }
+            };
+
             let amount = Amount::from(proof.amount);
 
             let secret_key = {
                 let keyset_cache_read_lock = self.keyset_cache.0.read().await;
 
-                let keyset = keyset_cache_read_lock
-                    .get(&keyset_id)
-                    .ok_or(Status::not_found(Error::KeysetNotFound(keyset_id)))?;
-                keyset
-                    .get(&amount)
-                    .ok_or(Status::not_found(Error::AmountNotFound(amount, keyset_id)))?
-                    .secret_key
-                    .clone()
+                let keyset = match keyset_cache_read_lock
+                .get(&keyset_id){
+                    Some(keyset) => keyset,
+                    None => {
+                        err_details.add_bad_request_violation(
+                            "keyset_id".to_string(),
+                            "keyset not found".to_string()
+                        );
+                        let status = Status::with_error_details( 
+                            Code::NotFound,
+                            "keyset not found".to_string(),
+                            err_details
+                        );
+                        return Err(status);
+                    }
+                };
+
+                match keyset.get(&amount){
+                    Some(key) => key.secret_key.clone(),
+                    None => {
+                        err_details.add_quota_failure_violation(
```

Comment:
https://docs.rs/tonic-richer-error/latest/tonic_richer_error/struct.ErrorDetails.html#method.add_bad_request_violation

Reply:
you didn't fixed that. use add_bad_request_violation instead